pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -707,11 +707,23 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
 			}
 
 			// Mount device to global mount path
-			err = volumeDeviceMounter.MountDevice(
+			operationState, err := volumeDeviceMounter.MountDevice(
 				volumeToMount.VolumeSpec,
 				devicePath,
 				deviceMountPath)
 			if err != nil {
+				switch operationState {
+				case volumetypes.OperationInProgress:
+					markDeviceUncertainError := actualStateOfWorld.MarkDeviceAsUncertain(volumeToMount.VolumeName, devicePath, deviceMountPath)
+					if markDeviceUncertainError != nil {
+						klog.Errorf(volumeToMount.GenerateErrorDetailed("MountDevice.MarkDeviceAsUncertain failed", markDeviceUncertainError).Error())
+					}
+				case volumetypes.OperationFinished:
+					markDeviceUnmountError := actualStateOfWorld.MarkDeviceAsUnmounted(volumeToMount.VolumeName)
```

Comment:
> The problem appears to be caused by remounting mechanism whereas a failing remount can mark a previously successful volume as "unmounted".

I think this is the place that broke Secrets in e2e tests.

Just an idea: with the new interface, Secrets, ConfigMaps, Projected and DownwardAPI volumes could return `NoChange` on Kube API errors and the volume status would not be changed.

Reply:
We could but I am currently thinking of keeping that behaviour out of those drivers and just for CSI. The fix I am making does it such that - if a volume was mounted previously, then next mount(re) failure should not mark the volume as unmounted. 

Each case where `NoChange` or such new change is added - we will have to carefully review that code.